SYNTHESIZER: Add counterexample-guided loop assigns synthesis

qinheping opened the pull request:
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.


Implement the functionality described below.

Motivation
---
This loop assigns synthesizer use the idea counter-example-guided synthesis (CEGIS) to synthesize loop assigns targets. It is based on the same CEGIS framework as the loop invariant synthesizer w ith the following difference.

Use Cases
---
Our loop invariants synthesizer will synthesize invariants clauses for all loops without loop invariant annotation. The loop assigns synthesizer will synthesize assigns clauses for all loops without loop invariant annotation **and** loop assigns annotation.

Cause Loop ID
---
In the loop invariant synthesizer, we say a loop is a cause loop if the violation is dependent on the loop's havoc instructions. Because invariant clauses affect the havoced values which could flow into variables outside the loop. In the loop assigns synthesizer, we say a loop is a cause loop if the violation is in the loop. Because assigns clauses only affect the CAR of its own loop body.

Synthesizer
---
For an assignable violation with checked target `T`, the new assigns target will be `T`. If `T` is a non-constant target, e.g., `ptr[i]` with `i` also an assigns target, we widen it to `__CPROVER_POINTER_OBJECT(T)`. We add the new target to cause loop one by one starting from the most-inner one. Until the assignable violation is resolved.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

jimgrundy commented on regression/goto-synthesizer/loop_contracts_synthesis_05/main.c:
```diff
@@ -0,0 +1,48 @@
+
+void main()
+{
+  struct blob *b = malloc(sizeof(struct blob));
```
Maybe insert some guards for the case where malloc returns NULL? This isn't default CBMC behavior now, but may become so in a subsequent major release (and will be disabled by an option rather than enabled).

jimgrundy commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1458,10 +1458,10 @@ void code_contractst::apply_loop_contracts(
 
       // Store loop number for
-      // ASSIGN ENTERED_LOOP = TRUE
+      // ASSIGN ENTERED_LOOP = true
```
Not sure what to make of these two comments, but they seem to say the opposite thing to eachother.

qinheping replied:
Yes, we store loop number for both of them. I clarify in the comment.

jimgrundy commented on src/goto-instrument/contracts/utils.cpp:
```diff
@@ -448,9 +448,19 @@ void generate_history_variables_initialization(
 
+bool is_transformed_loop_head(const goto_programt::const_targett &target)
+{
+  // The head of a transformed loop is
```
is this just

return is_assignment_to_instrumented_variable(target, ENTERED_LOOP) &&
           (target->assign_rhs() == false_exprt());

?

Seems like a more straightforward way to say it.

jimgrundy commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -93,11 +94,45 @@ optionst cegis_verifiert::get_options()
+    {
+      result.push_front(
+        loop_idt(step.function_id, original_loop_number_map[step.pc]));
+      continue;
```
Would the control flow be easier to follow if you didn't use "continue" and instead made the following "if" an "else if"?

jimgrundy commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -537,49 +580,89 @@ optionalt<cext> cegis_verifiert::verify()
-      checker->get_traces()[property_it.first], property_it.second.pc);
+    std::list<loop_idt> cause_loop_ids;
+    // Decide whether the violation is in the cause loop.
+    bool is_violation_in_loop;
```
I guess this defaults to false (I'm not much of a C++er)

jimgrundy replied:
Also looks like you could/should move this declaration after the next loop, because while you do set the value in the loop if you do you also return, so you don't use the set value. That seems a little confusing.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1457,11 +1457,11 @@ void code_contractst::apply_loop_contracts(
-      // ASSIGN ENTERED_LOOP = TRUE
+      // Store loop number for two type of instrumented instructions.
+      // ASSIGN ENTERED_LOOP = true
+      // ASSIGN ENTERED_LOOP = false
```
Further to what Jim said: perhaps you want to explain here that `ENTERED_LOOP = false` is assigned when leaving the loop?

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -708,9 +708,15 @@ void instrument_spec_assignst::inclusion_check_assertion(
 
-  dest.add(goto_programt::make_assertion(
+  goto_programt::instructiont to_add = goto_programt::make_assertion(
     inclusion_check_full(car, allow_null_lhs, include_stack_allocated),
-    source_location));
+    source_location);
+  // Record what target is checked.
+  auto &checked_assigns =
+    static_cast<exprt &>(to_add.condition_nonconst().add(ID_checked_assigns));
+  checked_assigns = car.target();
+
+  dest.add(std::move(to_add));
```
I'd suggest to first construct the condition of the assertion as a separate `exprt` (including the `ID_checked_assigns`), and then do `dest.add(goto_programt::make_assertion(modified_condition, source_location)`.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -93,11 +94,46 @@ optionst cegis_verifiert::get_options()
+    {
+      const loop_idt loop_id(
+        step.function_id, original_loop_number_map[step.pc]);
+      INVARIANT(result.front() == loop_id, "Leaving a loop we havn't entered.");
```
s/havn/haven/

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -537,49 +581,86 @@ optionalt<cext> cegis_verifiert::verify()
+    // Doing assigns-synthesis or invariant-synthesis
+    if(violation_type == cext::violation_typet::cex_assignable)
+    {
+      checked_pointer = static_cast<const exprt &>(
+        property_it.second.pc->condition().find(ID_checked_assigns));
+      cause_loop_ids = get_cause_loop_id_for_assigns(trace);
+      cext result(violation_type);
+      result.cause_loop_ids = cause_loop_ids;
+      result.checked_pointer = checked_pointer;
```
I'd put `cext result(violation_type);` and would then avoid the use of temporaries for you can safely do the assignments to `result.cause_loop_ids` and `result.checked_pointer` directly. I'd then suggest to inspect whether there is any use of `checked_pointer` and `cause_loop_ids` left, or whether these can be removed entirely. If they have to stay, then place the declarations as close as possible to the actual use.

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -106,11 +108,63 @@ void enumerative_loop_contracts_synthesizert::init_candidates()
+    new_assign.add_source_location() = source_location;
+
+    // Avoid adding same target.
+    if(assigns_map[loop_id].count(new_assign) == 0)
+    {
+      assigns_map[loop_id].insert(new_assign);
+      return;
+    }
```
Do just one lookup by using `if(assigns_map[loop_id].insert(new_assign).second) return;`

tautschnig commented on src/util/irep_ids.def:
```diff
@@ -330,6 +330,7 @@ IREP_ID_ONE(builtin_offsetof)
 IREP_ID_ONE(0)
 IREP_ID_ONE(1)
 IREP_ID_ONE(sizeof)
+IREP_ID_ONE(checked_assigns)
```
Nit pick: any particular reason for using this line? I'd suggest to just add it at the very end.

qinheping committed:
Add counterexample-guided loop assigns synthesis

Implement the functionality described below.

Motivation
---
This loop assigns synthesizer use the idea counter-example-guided synthesis (CEGIS) to synthesize loop assigns targets. It is based on the same CEGIS framework as the loop invariant synthesizer w ith the following difference.

Use Cases
---
Our loop invariants synthesizer will synthesize invariants clauses for all loops without loop invariant annotation. The loop assigns synthesizer will synthesize assigns clauses for all loops without loop invariant annotation **and** loop assigns annotation.

Cause Loop ID
---
In the loop invariant synthesizer, we say a loop is a cause loop if the violation is dependent on the loop's havoc instructions. Because invariant clauses affect the havoced values which could flow into variables outside the loop. In the loop assigns synthesizer, we say a loop is a cause loop if the violation is in the loop. Because assigns clauses only affect the CAR of its own loop body.

Synthesizer
---
For an assignable violation with checked target `T`, the new assigns target will be `T`. If `T` is a non-constant target, e.g., `ptr[i]` with `i` also an assigns target, we widen it to `__CPROVER_POINTER_OBJECT(T)`. We add the new target to cause loop one by one starting from the most-inner one. Until the assignable violation is resolved.
